### PR TITLE
controller: Fixed removal of apps with uuids as names

### PR DESCRIPTION
**Pull request by arekkas**
Follow up to #1467

Unit tests now require the uuid package. Not sure if I need to update godeps for that.

ping @lmars

### CODE REVIEW

**Comment by arekkas** [comment]
Seems like I need to update the godeps. However, neither `godep save` nor `godep save ./controller/` work :/

**Comment by jvatic** [comment]
@arekkas `godep save -r ./...` (you'll probably need to `godep restore ./...` first)

**Comment by arekkas** [comment]
Thanks :) Should I create a second commit for the godeps or amend it?

**Comment by jvatic** [comment]
A second commit for now, can always squash later

**Comment by arekkas** [comment]
Still not working:

```
vagrant@flynn:~/go/src/github.com/flynn/flynn$ godep restore ./...
package github.com/docker/docker/autogen/dockerversion
	imports github.com/docker/docker/autogen/dockerversion
	imports github.com/docker/docker/autogen/dockerversion: cannot find package "github.com/docker/docker/autogen/dockerversion" in any of:
	/usr/local/go/src/github.com/docker/docker/autogen/dockerversion (from $GOROOT)
	/home/vagrant/go/src/github.com/docker/docker/autogen/dockerversion (from $GOPATH)
godep: restore: exit status 1
package github.com/docker/docker/autogen/dockerversion
	imports github.com/docker/docker/autogen/dockerversion
	imports github.com/docker/docker/autogen/dockerversion: cannot find package "github.com/docker/docker/autogen/dockerversion" in any of:
	/usr/local/go/src/github.com/docker/docker/autogen/dockerversion (from $GOROOT)
	/home/vagrant/go/src/github.com/docker/docker/autogen/dockerversion (from $GOPATH)
godep: restore: exit status 1
[repeats...]
```

```
vagrant@flynn:~/go/src/github.com/flynn/flynn$ godep save -r ./...
godep: cannot find package "github.com/docker/docker/registry/v2" in any of:
	/usr/local/go/src/github.com/docker/docker/registry/v2 (from $GOROOT)
	/home/vagrant/go/src/github.com/docker/docker/registry/v2 (from $GOPATH)
godep: error loading dependencies
```

**Comment by jvatic** [comment]
You may need to manually `go get` the packages being reported by the save command. It will probably also make more changes than expected (i.e. remove a bunch of packages it doesn't think are needed). Godep is a bit flaky :/

**Comment by arekkas** [comment]
I can't seem to be able to go get `github.com/docker/docker/registry/v2`:

```
$ go get github.com/docker/docker/registry/v2
package github.com/docker/docker/registry/v2
	imports github.com/docker/docker/registry/v2
	imports github.com/docker/docker/registry/v2: cannot find package "github.com/docker/docker/registry/v2" in any of:
	/usr/local/go/src/github.com/docker/docker/registry/v2 (from $GOROOT)
	/home/vagrant/go/src/github.com/docker/docker/registry/v2 (from $GOPATH)
```

**Comment by jvatic** [comment]
Interesting, cloning `github.com/docker/docker` into `$GOPATH/src/github.com/src/docker/docker` should get around that.

**Comment by arekkas** [comment]
Nope, doesn't work either. `autogen` does not exist in the repository, maybe some script needs to be executed first?

**Comment by jvatic** [comment]
Try running `go generate` in `.../docker/registry`.

**Comment by arekkas** [comment]
doesn't generate the "autogen" directory either. Also tried `go generate ./...` in `$gopath.../docker`

**Comment by arekkas** [comment]
```
export DOCKER_CLIENTONLY=1
chmod +x hack/make/.go-autogen && hack/make/.go-autogen
```

did the trick for autogen, "github.com/docker/docker/registry/v2" still doesn't work though.

**Comment by jvatic** [comment]
Hmm, `git reset --hard 8295bed` from within `$GOPATH/.../docker` and `cd registry; go generate`?

**Comment by arekkas** [comment]
Doesn't help either :/

I got a little further:

```
$ rm $GOPATH/.../docker
$ go get github.com/docker/docker/...
$ cd $GOPATH/.../docker
$ git reset --hard 8295bed
$ export DOCKER_CLIENTONLY=1
$ chmod +x hack/make/.go-autogen && hack/make/.go-autogen
$ godep restore ./...
$ godep save -r ./...
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/api
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/daemon/graphdriver
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/daemon/networkdriver
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/engine
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/graph
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/image
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/nat
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/opts
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/archive
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/chrootarchive
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/devicemapper
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/directory
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/fileutils
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/homedir
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/httputils
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/ioutils
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/mflag
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/mount
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/parsers
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/pools
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/progressreader
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/promise
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/reexec
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/signal
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/stringid
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/stringutils
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/system
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/tarsum
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/term
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/timeoutconn
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/timeutils
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/truncindex
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/ulimit
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/units
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/pkg/version
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/registry
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/runconfig
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/utils
godep: dirty working tree: /home/vagrant/go/src/github.com/docker/docker/vendor/src/code.google.com/p/go/src/pkg/archive/tar
godep: error loading dependencies

```

**Comment by jvatic** [comment]
Looks like you'll need to do a `git add --all; git reset --hard` from within `$GOPATH/.../docker`

**Comment by arekkas** [comment]
That did it :) For future references:

```
$ rm -rf $GOPATH/src/github.com/docker/docker
$ go get github.com/docker/docker/...
$ cd $GOPATH/src/github.com/docker/docker
$ git reset --hard 8295bed
$ export DOCKER_CLIENTONLY=1
$ chmod +x hack/make/.go-autogen && hack/make/.go-autogen
$ git add --all; git reset --hard
$ godep restore ./...
$ godep save -r ./...
```

**Comment by arekkas** [comment]
Damn, unit tests still failing...

**Comment by jvatic** [comment]
Looks like there's a lot of unwanted deletes from `godep save` and it doesn't appear the new dependency was added.

**Comment by arekkas** [comment]
The uuid package is only needed for `go test`, maybe that's why it didn't get added?

**Comment by arekkas** [comment]
Do you have a custom uuid generator for flynn? Maybe I could use that one instead of google's uuid package so there wouldn't be the need for updating godeps

**Comment by lmars** [comment]
@arekkas https://github.com/flynn/flynn/blob/master/pkg/random/random.go#L37

**Comment by lmars** [comment]
P.S. Updating Godeps is very fragile, we really need to deal with that soon.

**Comment by arekkas** [comment]
I rolled back the godeps update and used `random.UUID()` instead.

I think that this would have worked though:

```
$ rm -rf $GOPATH/src/github.com/docker/docker
$ go get github.com/docker/docker/...
$ cd $GOPATH/src/github.com/docker/docker
$ git reset --hard 8295bed
$ export DOCKER_CLIENTONLY=1
$ chmod +x hack/make/.go-autogen && hack/make/.go-autogen
$ git add --all; git reset --hard
$ cd $GOPATH/src/github.com/flynn/flynn
$ go get ./...
$ godep restore ./...
$ godep save -r ./...
```

**Comment by arekkas** [comment]
Ok, tests are passing :) ready for :eyes: :)

File: controller/controller_test.go [hunk 1]
```diff
@@ -173,6 +173,36 @@ func (s *S) TestDeleteApp(c *C) {
...
+func (s *S) TestDeleteNonExistingApp(c *C) {
+	for _, useUuid := range []bool{false, true} {
+		var appID string
+
+		if useUuid {
+			appID = "foobar"
+		} else {
+			appID = random.UUID()
+		}
+		c.Assert(s.c.DeleteApp(appID), NotNil)
```
Comment (lmars) [inline controller/controller_test.go]: This should assert that the error is `controller.ErrNotFound` (and modify the code to return that if it doesn't already).

File: controller/controller_test.go [hunk 2]
```diff
@@ -173,6 +173,36 @@ func (s *S) TestDeleteApp(c *C) {
...
+			appID = app.ID
+		}
+		c.Assert(s.c.DeleteApp(appID), IsNil)
+
+		_, err := s.c.GetApp(appID)
+		c.Assert(err, Equals, controller.ErrNotFound)
+	}
+}
+
+func (s *S) TestDeleteNonExistingApp(c *C) {
```
Comment (lmars) [inline controller/controller_test.go]: I think `TestDeleteNonExistentApp` is a better name

File: controller/controller_test.go [hunk 3]
```diff
@@ -173,6 +173,36 @@ func (s *S) TestDeleteApp(c *C) {
...
+	for _, useUuid := range []bool{false, true} {
+		var appID string
+
+		if useUuid {
```
Comment (lmars) [inline controller/controller_test.go]: This should be `useUUID`

**Comment by arekkas** [comment]
Not sure why tests fail but it doesn't seem to be related to this PR. `go test ./controller` works fine :)

**Comment by lmars** [comment]
Thanks!

**Comment by arekkas** [comment]
Sure :) One last thing:
If I do `flynn install` now, is this change going to be included?

**Comment by lmars** [comment]
We release nightly at 6AM UTC, so that is when it will be available (provided master is passing CI at that time), but if you really need it let us know and we can push a manual release.

**Comment by arekkas** [comment]
No it's fine, thanks :)

**Pull request merged**